source/common/router/retry_state_impl.cc

```diff
@@ -265,5 +240,48 @@ bool RetryStateImpl::wouldRetry(const Http::HeaderMap* response_headers,
   return false;
 }
 
+bool RetryStateImpl::wouldRetryFromReset(const Http::StreamResetReason& reset_reason) {
+  if ((retry_on_ & RetryPolicy::RETRY_ON_REFUSED_STREAM) &&
+      reset_reason == Http::StreamResetReason::RemoteRefusedStreamReset) {
+    return true;
+  }
+
+  if ((retry_on_ & RetryPolicy::RETRY_ON_CONNECT_FAILURE) &&
+      reset_reason == Http::StreamResetReason::ConnectionFailure) {
+    return true;
+  }
+
+  return false;
+}
+
+bool RetryStateImpl::wouldRetry(const Http::HeaderMap* response_headers,
+                                const absl::optional<Http::StreamResetReason>& reset_reason) {
+  // First check "never retry" conditions so we can short circuit, then delegate to
+  // helper methods for checks dependant on retry policy.
+
+  // we never retry if the reset reason is overflow.
+  if (reset_reason && reset_reason.value() == Http::StreamResetReason::Overflow) {
+    return false;
+  }
+
+  if (response_headers != nullptr) {
+    // We never retry if the overloaded header is set.
+    if (response_headers->EnvoyOverloaded() != nullptr) {
+      return false;
+    }
+
+    if (wouldRetryFromHeaders(*response_headers)) {
+      return true;
+    }
+  } else if (retry_on_ & (RetryPolicy::RETRY_ON_5XX | RetryPolicy::RETRY_ON_GATEWAY_ERROR)) {
+    // wouldRetry() is passed null headers when there was an upstream reset. Currently we count an
```

Comment:
I think this comment should be removed from the place where you copied it from?

Reply:
Ah, yeah. Removed.